src/managed/Microsoft.DotNet.Build.Bundle/Bundler.cs

```diff
@@ -0,0 +1,205 @@
+using System.Reflection.PortableExecutable;
+
+namespace Microsoft.DotNet.Build.Bundle
+{
+    /// <summary>
+    /// Bundler: Functionality to embed the managed app and its dependencies
+    /// into the host native binary.
+    /// </summary>
+    public class Bundler
+    {
+        string HostName;
+        string ContentDir;
+        string OutputDir;
+        bool EmbedPDBs;
+
+        string Application;
+        string DepsJson;
+        string RuntimeConfigJson;
+        string RuntimeConfigDevJson;
+
+        /// <summary>
+        /// Align embedded assemblies such that they can be loaded 
+        /// directly from memory-mapped bundle.
+        /// TBD: Set the correct value of alignment while working on 
+        /// the runtime changes to load the embedded assemblies.
+        /// </summary>
+        const int AssemblyAlignment = 16;
+
+        public static string Version => (Manifest.MajorVersion + "." + Manifest.MinorVersion);
+
+        public Bundler(string hostName, string contentDir, string outputDir, bool embedPDBs)
+        {
+            ContentDir = contentDir;
+            OutputDir = outputDir;
+            HostName = hostName;
+            EmbedPDBs = embedPDBs;
+        }
+
+        void ValidateFiles()
+        {
+            // Check required directories
+            if (!Directory.Exists(ContentDir))
+            {
+                throw new BundleException("Dirctory not found: " + ContentDir);
+            }
+            if (!Directory.Exists(OutputDir))
+            {
+                throw new BundleException("Dirctory not found: " + OutputDir);
+            }
+
+            // Set default names
+            string baseName = Path.GetFileNameWithoutExtension(HostName);
+            Application = baseName + ".dll";
+            DepsJson = baseName + ".deps.json";
+            RuntimeConfigJson = baseName + ".runtimeconfig.json";
+            RuntimeConfigDevJson = baseName + ".runtimeconfig.dev.json";
+
+            // Check that required files exist on disk.
+            Action<string> checkFileExists = (string name) =>
+            {
+                string path = Path.Combine(ContentDir, name);
+                if (!File.Exists(path))
+                {
+                    throw new BundleException("File not found: " + path);
+                }
+            };
+
+            checkFileExists(HostName);
+            checkFileExists(Application);
+            // The *.json files may or may not exist.
+        }
+
+        /// <summary>
+        /// Embed 'file' into 'bundle'
+        /// </summary>
+        /// <returns>Returns the offset of the start 'file' within 'bundle'</returns>
+
+        long AddToBundle(Stream bundle, Stream file, FileType type = FileType.Extract)
+        {
+            // Allign assemblies, since they are loaded directly from bundle
+            if (type == FileType.Assembly)
+            {
+                long padding = AssemblyAlignment - (bundle.Position % AssemblyAlignment);
```

Comment:
`padding` will equal `AssemblyAlignment` if no alignment is needed, wasting `AssemblyAlignment` bytes.